chia/wallet/wallet_node.py

```diff
@@ -1028,180 +1027,181 @@ async def new_peak_wallet(self, new_peak: wallet_protocol.NewPeakWallet, peer: W
-                            height=peer_new_peak_height,
-                            header_hash=peer_new_peak_hash,
-                        )
-                        if success:
-                            self.synced_peers.add(peer.peer_node_id)
-                    else:
-                        if peak_hb is not None and new_peak.weight <= peak_hb.weight:
-                            # Don't process blocks at the same weight
-                            return
+        if new_peak_hb.height < self.constants.WEIGHT_PROOF_RECENT_BLOCKS:
+            # this is the case happens chain is shorter then WEIGHT_PROOF_RECENT_BLOCKS
+            await self.sync_from_untrusted_close_to_peak(new_peak_hb, peer)
+            return
 
-                    # For every block, we need to apply the cache from race_cache
-                    for potential_height in range(backtrack_fork_height + 1, new_peak.height + 1):
-                        header_hash = self.wallet_state_manager.blockchain.height_to_hash(uint32(potential_height))
-                        if header_hash in self.race_cache:
-                            self.log.info(f"Receiving race state: {self.race_cache[header_hash]}")
-                            await self.receive_state_from_peer(list(self.race_cache[header_hash]), peer)
+        if not far_behind and peer.peer_node_id in self.synced_peers:
+            # This is the (untrusted) case where we already synced and are not too far behind. Here we just
+            # fetch one by one.
+            await self.sync_from_untrusted_close_to_peak(new_peak_hb, peer)
+            return
 
-                    self.wallet_state_manager.state_changed("new_block")
-                    self.wallet_state_manager.set_sync_mode(False)
-                    self.log.info(f"Finished processing new peak of {new_peak.height}")
+        # we haven't synced fully to this peer yet
+        syncing = False
+        if far_behind or len(self.synced_peers) == 0:
+            syncing = True
+            self.wallet_state_manager.set_sync_mode(True)
 
-        if peer.peer_node_id in self.synced_peers:
-            await self.wallet_state_manager.blockchain.set_finished_sync_up_to(new_peak.height)
+        secondery_sync_running = (
+            self._secondary_peer_sync_task is not None and self._secondary_peer_sync_task.done() is False
+        )
+        if not syncing and secondery_sync_running:
+            self.log.info("Will not do secondary sync, there is already another sync task running.")
+            return
+
+        if await self.check_for_synced_trusted_peer(new_peak_hb, request_time):
+            self.log.info("Cancelling untrusted sync, we are connected to a trusted peer")
+            return
+
+        try:
+            await self.long_sync_from_untrusted(syncing, new_peak_hb, peer)
+        except Exception as e:
+            tb = traceback.format_exc()
+            self.log.error(f"Error syncing to {peer.get_peer_info()} {e} {tb}")
```

Comment:
using `log.exception(f"Error syncing to {peer.get_peer_info()}")` will automatically add the stacktrace